pyccel/ast/cwrapper.py

```diff
@@ -491,14 +524,63 @@ def pybool_to_bool(cast_function_name):
     'pybool_to_bool' : pybool_to_bool,
 }
 
+
+PyArray_CheckScalar = FunctionDef(name = 'PyArray_CheckScalar',
+                                  body= [],
+                                  arguments = [Variable(dtype=PyccelPyObject(), name = 'o', is_pointer=True)],
+                                  results = [Variable(dtype=NativeBool(), name = 'r')])
+
+PyArray_ScalarAsCtype = FunctionDef(name = 'PyArray_ScalarAsCtype',
+                                    body = [],
+                                    arguments = [Variable(dtype=PyccelPyObject(), name = 'o', is_pointer=True),
+                                                Variable(dtype=NativeVoid(), name = 'c', is_pointer = True)],
+                                    results = [])
+
 collect_function_registry = {
     NativeInteger(): PyLong_AsLong,
     NativeReal() : PyFloat_AsDouble,
 }
 
-check_type_registry = {
+check_type_registry  = {
     NativeInteger(): 'PyLong_Check',
     NativeComplex() : 'PyComplex_Check',
     NativeReal() : 'PyFloat_Check',
     NativeBool() : 'PyBool_Check',
 }
+
+
+
+# Needed to check for numpy arguments type
+Numpy_Bool_ref = Variable(dtype=NativeVoid(),  name = 'Bool')
+Numpy_Byte_ref = Variable(dtype=NativeVoid(),  name = 'Byte')
+Numpy_Short_ref = Variable(dtype=NativeVoid(),  name = 'Short')
+Numpy_Int_ref = Variable(dtype=NativeVoid(),  name = 'Int')
+Numpy_Long_ref = Variable(dtype=NativeVoid(),  name = 'Long')
```

Comment:
I can't find where these values are defined. Is there an equivalent variable with the precision in the name? Numpy contains both `np.long` and `np.int64`. See:
https://numpy.org/doc/stable/user/basics.types.html

I'm not certain that a `long` has the same precision on linux and windows. If an `Int64` exists in numpy and this fixes your type collection problems on windows then we also need to update the type that we use in our code to be certain that it matches, here:
https://github.com/pyccel/pyccel/blob/master/pyccel/codegen/printing/ccode.py#L194-L202
And probably in the array code too.
@saniQQ can you say what exactly would need updating

Reply:
Here is a link with a better explanation of what I think the problem is:
https://blogs.igalia.com/mshin/2019/08/26/why-should-we-use-a-precise-width-integer-type/

> Data model | sizeof(int) | sizeof(long) | sizeof(long long) | example
> -- | -- | -- | -- | --
> LP32 | 16b | 32b | 64b | Win16
> ILP32 | 32b | 32b | 64b | Win32, i386 OSX & Linux
> LP64 | 32b | 64b | 64b | x86-64 OSX & Linux
> LLP64 | 32b | 32b | 64b | Win64

Reply:
Yes I know the problem when using PyArray_IsScalar(pyobject,  type)     when giving Long as type it fails on windows because windows64 long is 32bits size so I tried using LongLong in type it works in windows and it fails on Linux ...

Reply:
Does this mean that there is no `Int64` type provided by numpy?

Reply:
The solution is not to change the imprecise types we use to other imprecise types as these will indeed be prone to the same problems. My suggested solution is to instead use precise types (i.e. `int8_t`, `int16_t`, `int32_t`, `int64_t`) see:
https://en.cppreference.com/w/c/types/integer

Reply:
To explain better the problem stop at the check type stage so it comes from PyArray_IsScalar(pyobject, type) return false when it should return true.

Reply:
we will need to update :
- the union in the t_ndarray struct 
- the enum (this will need to update the dictionary in ccode.py too)
- the Generic Macro to use the precise types 
- and the fill functions

Reply:
This will affect PR #562 as @rakati added the fill functions to `ndarrays.h` and to the Generic Macro

Reply:
i can push the changes here so @rakati can merge them later

Reply:
That sounds like a good suggestion :+1: